hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationTimer.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.partition.impl;
+
+import com.hazelcast.internal.util.Timer;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.LongAccumulator;
+
+/**
+ * the total elapsed time of executions from source to destination endpoints
+ */
+public class MigrationTimer {
+    /**
+     * on the latest repartitioning round.
+     */
+    private final LongAccumulator elapsedNanoseconds = new LongAccumulator(Long::max, 0);
```

Comment:
This `LongAccumulator` is only updated by receiving the value of `volatile long lastRepartitionNanos` after repartitions complete - is there a particular circumstance where it's possible for this field to receive conflicting values, and thus need an accumulator to select the max passed value? If not, it's probably fine to just use a volatile long here as well.

Reply:
> This `LongAccumulator` is only updated by receiving the value of `volatile long lastRepartitionNanos` after repartitions complete - is there a particular circumstance where it's possible for this field to receive conflicting values, and thus need an accumulator to select the max passed value? If not, it's probably fine to just use a volatile long here as well.

I'm not quite sure I follow:

are you asking if the `max` is necessary as a pose to just taking the most recent value?
If so then that was the way it was implemented before, but there was a race condition where an earlier update could be blocked and then resume later, replacing a later value that'd already been processed.

If your asking if I could do the `max` operation with a `volatile long` - I couldn't think of an easy way to do that atomically.

Either way - yes I think it's needed?

Reply:
>> are you asking if the `max` is necessary as a pose to just taking the most recent value? If so then that was the way it was implemented before, but there was a race condition where an earlier update could be blocked and then resume later, replacing a later value that'd already been processed.

Looking again at the code, because it's now such a small part and in one line it's actually quite difficult to see what the problem was before.
Before it was something like
```
final long newElapsed = Timer.nanosElapsed(lastRepartitionNanos.get());

< do something else >

totalElapsedTime.doSomethingWith(newElapsed);
```

The code was written assuming that `newElapsed` would always be the most recent, but that isn't _always_ the case.

Reply:
Understood, I was just asking if it was possible for a race condition to occur, which you've confirmed - so sounds like the right solution to me 👍